refactor (#334)
KxSystems/docs: docs/interfaces/q-server-for-odbc3.md

# :fontawesome-solid-database: kdb+ server for ODBC3

> The ODBC3 driver allows applications to query kdb+ via the ODBC interface.  

Reviewer: ```suggestion
The ODBC3 driver allows applications to query kdb+ through the ODBC interface.  
```

!!! tip "Reporting problems"

    When reporting a problem (e.g. SQL error, wrong results, slowness, segfault etc.) make sure to include steps to reproduce along with your [ODBC trace](#debugging).

:fontawesome-regular-map: [**Data visualization with kdb+ using ODBC**: a Tableau case study](../wp/data-visualization/index.md "White paper")

## Installation

### kdb+ configuration

The kdb+ process should be [listening on port](../basics/listening-port.md) which relates to the port choosen and defined in the odbc configuration.

Unzip qodbc3.zip, and copy `ps.k` to `$QHOME`. Ensure you have `ps.k` loaded into the kdb+ process:

```q
q)\l ps.k
```

Run `.s.ver` to check that ps.k has been loaded and for the current version e.g.
```q
q).s.ver
1.15
```

> The following is an example of staring kdb+ with ps.k listening on port 5000

Reviewer: ```suggestion
The following is an example of starting kdb+ with ps.k listening on port 5000
```

```bash
q ps.k -p 5000
```

### ODBC driver and DSN configuration

> Software that uses ODBC connections refer to individual connections via a `DSN`(Data Source Name). 

Reviewer: ```suggestion
Software that uses ODBC connections refer to individual connections using a `DSN`(Data Source Name). 
```

$ ld -o qodbc3.so -shared qodbc3.o c.o -lodbc -lodbcinst -lm
```

The ODBC driver should be installed in the corresponding system `ini` file and a DSN entry created in either the system data sources (for all users) or user data source (current user).
The linux `odbcinst` tool can be used to find these locations if not known, for example:

```bash
$ odbcinst -j
unixODBC 2.3.7
DRIVERS............: /etc/odbcinst.ini
SYSTEM DATA SOURCES: /etc/odbc.ini
FILE DATA SOURCES..: /etc/ODBCDataSources
USER DATA SOURCES..: /root/.odbc.ini
SQLULEN Size.......: 8
SQLLEN Size........: 8
SQLSETPOSIROW Size.: 8
```

> Add the driver to your list of drivers by editing the driver ini file ( `/etc/odbcinst.ini`). An example of the definition follows (note: the driver locaton should be altered to the location choosen on your system):

Reviewer: ```suggestion
Add the driver to your list of drivers by editing the driver `ini` file ( `/etc/odbcinst.ini`). An example of the definition follows (note: the driver location should be altered to the location chosen on your system):
```

Also, SQL selects from partitioned tables are not supported – one should pre-select from a partitioned table using the `q()` function instead.

`test.q` provides additional examples of SQL usage, including the create/insert/update/delete statement syntax.

## Debugging

ODBC implementations provide a tracing capability to log interactions with an ODBC driver. This can aid in diagnosing any issues. Tracing can have a detremental 
effect to ODBC performance.

> This is a feature of the ODBC system, and not a feature unique to the kdb+ driver. Associated documentation, features and troubleshooting can be found online for your OS.

Reviewer: ```suggestion
This is a feature of the ODBC system, and not a feature unique to the kdb+ driver. Associated documentation, features, and troubleshooting can be found online for your OS.
```

!!! warn "Connection details can be logged, therefore the log file should be stored in a private location"

### Windows

In the ODBC data source administrator, click _Start Tracing_ on the _Tracing_ tab. 

> An example of the data recorded to the `SQL.LOG`:

Reviewer: ```suggestion
See below an example of the data recorded to the `SQL.LOG`:
```

```txt
powershell      da8-9a0	ENTER SQLExecDirectW
		        HSTMT               0x00000256C37A6920
                WCHAR *             0x00000256AA98112C [      -3] "SELECT * FROM t\ 0"
                SDWORD                    -3
...
powershell      da8-9a0	EXIT  SQLGetData  with return code 0 (SQL_SUCCESS)
		        HSTMT               0x00000256C37A6920
		        UWORD                        2
		        SWORD                       -8 <SQL_C_WCHAR>
		        PTR                 0x00000256A79BB1D0 [       2] "1"
		        SQLLEN                  4092
		        SQLLEN *            0x000000DAA198E160 (2)
```

Please ensure tracing is disabled after debugging complete.

### Linux

> In the `[ODBC]` section of the driver ini file (`/etc/odbcinst.ini`), set `Trace` to `1` to enable tracing and `TraceFile` to the location of the log file to create. 

Reviewer: ```suggestion
In the `[ODBC]` section of the driver `ini` file (`/etc/odbcinst.ini`), set `Trace` to `1` to enable tracing and `TraceFile` to the location of the log file to create. 
```
